rsh: Insert TERM env variable into the command to be run

jwendell opened the pull request:
Fixes the CLI part of #7424.

liggitt commented on pkg/cmd/cli/cmd/rsh.go:
```diff
@@ -143,5 +144,15 @@ func (o *RshOptions) Validate() error {
+	}
+
+	// If it failed, try again the original command, without the TERM addition
+	o.Command = originalCommand
```
This seems super sketchy. Any rsh session returning a failure is run twice? It also assumes all commands are ok to run twice?

jwendell replied:
Indeed, this is a bad fix... I'll think on another approach

liggitt commented on pkg/cmd/cli/cmd/rsh.go:
```diff
@@ -143,5 +144,15 @@ func (o *RshOptions) Validate() error {
 func (o *RshOptions) Run() error {
+	// Insert the TERM into the command to be run
+	originalCommand := o.Command
+	o.Command = []string{"/bin/sh", "-c", fmt.Sprintf("TERM=vt100 %s", strings.Join(o.Command, " "))}
```
also, we can't assume that joining args with spaces is correct... this doesn't handle quoting/escaping

jwendell commented:
Now with one-line patch. Borrowed the idea from https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockertools/exec.go#L48

jwendell commented:
@liggitt still didn't like the approach?

liggitt commented:
it's a lot better. @ncdc or @rhcarvalho want to weigh in?

ncdc commented:
Why not use the current shell's $TERM instead of hardcoding to vt100?

ncdc commented:
Also please note that this isn't 100% useful until my PR that adds support for terminal resizing lands

ncdc commented:
Reviewers encouraged 😄 https://github.com/kubernetes/kubernetes/pull/25273

jwendell commented:
@ncdc Now it's defaulting to current TERM's variable ;)

rhcarvalho commented:
@ncdc while having resizing support will be awesome, I think this will be useful even before that!
For example, the issue description (https://github.com/openshift/origin/issues/7424) depicts the current situation when running `top` in a container:

```console
$ oc rsh docker-registry-1-x5d7z
bash-4.2$ top
TERM environment variable not set.
```

rhcarvalho commented on pkg/cmd/cli/cmd/rsh.go:
```diff
@@ -143,5 +144,10 @@ func (o *RshOptions) Validate() error {
 // Run starts a remote shell session on the server
 func (o *RshOptions) Run() error {
+	// Insert the TERM into the command to be run
+	// TODO what if the container doesn't have `env`???
```
I assume most usages of `oc rsh` use the default command (`/bin/sh`) -- "rsh" standing for remote shell. Minimal operating systems like Busybox, and Alpine include both `/bin/sh` and `/bin/env`.

For more control, one can always use `oc exec`.

rhcarvalho commented on pkg/cmd/cli/cmd/rsh.go:
```diff
@@ -143,5 +144,10 @@ func (o *RshOptions) Validate() error {
 func (o *RshOptions) Run() error {
+	// Insert the TERM into the command to be run
+	// TODO what if the container doesn't have `env`???
+	term := fmt.Sprintf("TERM=%s", util.Env("TERM", "vt100"))
```
If there is to be any default, I'd go with the same value that Docker uses for `docker run`: xterm.
[docker/docker/container/container_unix.go#L61-L63](https://github.com/docker/docker/blob/cbf785c45a8cca061028ec341320bcbd655cf1fd/container/container_unix.go#L61-L63)

rhcarvalho commented:
Missing:

- [ ] Tests
- [ ] Support for `oc exec -t`

rhcarvalho commented:
I think the implementation is in the wrong file.
`oc rsh` has flags to force enable or force disable TTY allocation.

We should only set TERM when we have TTY=true.

It would be "easier" to change code in `Godeps/_workspace/src/k8s.io/kubernetes/pkg/kubectl/cmd/exec.go`, which means fixing the problem upstream...

rhcarvalho commented:
I've opened this issue upstream https://github.com/kubernetes/kubernetes/issues/28280, if it gets solved upstream for `kubectl exec`, our `oc rsh` and `oc exec -t` will inherit the fix.

jwendell commented:
@rhcarvalho It seems there are no tests for rsh at all... wrt oc exec, it's just a wrapper for kubectl. If we touch it, it should be made upstream, and tbh I don't know they are like it (see the PR I sent).

@ncdc Now I've updated docs to say we send a TERM variable; And changed fallback value to xterm instead of vt100.

ncdc commented:
@jwendell for now we only need the fix in `oc rsh`

jwendell commented:
@ncdc, ok. this patch only touches oc rsh...

rhcarvalho commented on pkg/cmd/cli/cmd/rsh.go:
```diff
@@ -24,9 +25,11 @@ This command will attempt to start a shell session in a pod for the specified re
-if standard input is interactive - use -t and -T to override.
+if standard input is interactive - use -t and -T to override. A TERM variable is sent
+to the environment where the shell (or command) will be executed. By default its value
+is the same as the TERM variable from the local environment; If not found, 'xterm' is used.
```
s/If/if ?

Maybe "if not set" would be more precise?

rhcarvalho commented:
The current implementation is fine by me, @jwendell  sorry I don't know whether we have tests for `oc rsh`.

jwendell commented:
So, what's necessary to have this merged? Only this minor wording thing? Anything else?

fabianofranz commented:
@jwendell yes, update that minor nit and we'll get this tagged.

fabianofranz commented:
[test]

fabianofranz commented:
[merge]
